master/src/main/java/org/apache/celeborn/service/deploy/master/SlotsAllocator.java

```diff
@@ -238,6 +253,23 @@ private static List<Integer> roundRobin(
     return partitionIdList;
   }
 
+  private static boolean haveUsableSlots(
+      Map<WorkerInfo, List<UsableDiskInfo>> restrictions, List<WorkerInfo> workers, int index) {
+    return restrictions.get(workers.get(index)).stream().mapToLong(i -> i.usableSlots).sum() > 0;
+  }
+
+  private static boolean satisfyRackAware(
+      CelebornRackResolver rackResolver,
+      List<WorkerInfo> workers,
+      int masterIndex,
+      int nextSlaveInd) {
+    return rackResolver == null
+        || !rackResolver.enabledRackAware()
+        || rackResolver.getDistance(
```

Comment:
I think we can just check if the two nodes have the same parent, like ```NetworkTopology.isOnSameRack``` in hadoop 3.2.1
```
  public boolean isOnSameRack( Node node1,  Node node2) {
    if (node1 == null || node2 == null) {
      return false;
    }
      
    netlock.readLock().lock();
    try {
      return isSameParents(node1, node2);
    } finally {
      netlock.readLock().unlock();
    }
  }
```

Reply:
Done